src/internals/job_lifecycle.md

```diff
@@ -0,0 +1,199 @@
+# Job Lifecycle
+
+Treadmill jobs represent units of work schedulable on a supervisor. Each job is
+eligible to run on a set of supervisors, limited by a set of *tag filter
+expresssions* governed by the job request, and the permissions of the user who
+scheduled the job.
+
+From the point of their creation up to their successful completion or failure,
+jobs go through a set of *state changes*. A job's state is composed of two
+components: its *execution state*, and its *exit status*.
+
+A job's *execution state* describes the *current* state of the schedulable unit
+of work and is controlled by the Switchboard (e.g., by assigning a job on a
+particular supervisor), and the Supervisor (e.g., by starting or terminating a
+Virtual Machine). A job may not necessarily transition through all defined
+execution states; states may be arbitrarily skipped. All jobs must eventually
+end up in the `Terminated` state. **Only some transitions between different
+execution states are legal.** When attempting to take an illegal transition
+towards another execution state, the Switchboard may ignore this transition, or
+attempt to terminate the given job. A job that reached the `Terminated` state
+must not transition to other states; **a `Terminated` execution state is
+final**.
+
+The *exit status* is controlled by the supervisor and describes the user-visible
+state that the job assumes once it reaches the `Terminated` execution state. A
+job's exit status may be set multiple times, but must not change once the job's
+execution state reaches `Terminated`. The execution state can be used to
+communicate whether the Treadmill system was able to successfully schedule the
+job, whether there were any Treadmill system-internal errors that prevented its
+successful execution, and whether the user-defined job workload reported a
+success or error result. **Only some transitions between different exit statuses
+are legal.** When attempting to take an illegal transition towards another exit
+status, the previous exit status remains valid.
+
+We describe these two components as Rust enums below.
+
+## Execution State
+
+```rust
+enum ExecutionState {
+    /// A job object has been created, but it has not been assigned to a
+    /// supervisor yet.
+    ///
+    /// This is the starting state for newly created jobs.
+    Queued,
+
+    /// A job object has been assigned to a particular supervisor.
+    Scheduled,
```

Comment:
Not super clear to me what exactly makes `Scheduled` different from `Initializing`; my assumption when first reading was that `Scheduled` indicated that the switchboard hadn't yet sent a start message?

Reply:
So the idea was that we can only transition into `Initializing` by receiving such a state change from the Supervisor. However, `Scheduled` would mean that the Switchboard has begun starting a job (i.e., assigned it to a Supervisor), but that Supervisor has not yet confirmed that the job was picked up. It's a way for the `Switchboard` to transition a job out of `Queued` atomically, without contacting the Supervisor. I hoped that this would be useful, e.g., to distinguish jobs are still schedulable from those that are not.

I hope that the text makes it clear that not all of these states are mandatory to be implemented. Given that *only* the Switchboard can transition into this state, and it is allowed to skip this state, it is legal for it to not implement it. It may still be useful to have it defined in the API, such that clients understand it.

However, if you think that we'll never need this or it's redundant, we should remove it. Hope this clarifies things!

Reply:
Ah, I see now. Yes, that does make sense. The assumption I *was* using was that if the supervisor connection was still open, then the supervisor could be expected to be working, and thus fire-and-forget was mostly sufficient, but what you're proposing makes more sense imo.